Feat: send mail body

adrianmoloce opened the pull request:
- Created default template for the email message body
  - Created environment configuration for the template used
  - Created unit tests for configuring and rendering the body to check it being sent as expected.

amoloce committed:
feat: Add message body to the mail notification

  - Test if sendmail is called with expected body

amoloce committed:
feat: Create email body templates and configure by path.

  - Created email templates directory with default and test templates
  - Tested that configuring and rendering the templates sends the
    expected body
  - Created env variable for body template path and default values

koosie0507 commented on tests/test_web.py:
```diff
@@ -199,3 +200,46 @@ def test_smtp_starttls_env_flag_must_be_true_boolean(environment, http, server,
+        autoescape=select_autoescape(['html', 'xml'])
+    )
+    template = env.get_template(environment.get("INVOICE_UTILS_BODY_TEMPLATE_PATH"))
+    expected_body = template.render(
```
Extract code to a fixture to improve test readability

koosie0507 commented on tests/test_web.py:
The test template should be loaded from a subdirectory of the `tests` directory.

koosie0507 commented on tests/test_web.py:
```diff
@@ -199,3 +200,46 @@ def test_smtp_starttls_env_flag_must_be_true_boolean(environment, http, server,
+):
+    http.post("/invoice", json=email_invoice_request_body)
+
+    env = Environment(
```
The test knows too much about the template's logic. Ideally, the test just calls the endpoint and if a template is specified, the e-mail contains an expected string. There shouldn't be any template manipulation in the test.

amoloce committed:
fix: Extract test code into a fixture

  - Test doesn't contain code logic anymore
  - Created env vars to make the template location more configurable

koosie0507 commented on src/invoice_utils/config.py:
```diff
@@ -16,3 +19,9 @@ def _str_to_bool(arg: str) -> bool:
 INVOICE_UTILS_SMTP_TLS = _str_to_bool(os.getenv("INVOICE_UTILS_SMTP_TLS"))
+INVOICE_UTILS_BODY_TEMPLATE_NAME = os.getenv("INVOICE_UTILS_BODY_TEMPLATE_NAME", DEFAULT_BODY_TEMPLATE_NAME)
+INVOICE_UTILS_BODY_TEMPLATE_PACKAGE = os.getenv("INVOICE_UTILS_BODY_TEMPLATE_PACKAGE", DEFAULT_BODY_TEMPLATE_PACKAGE)
+INVOICE_UTILS_BODY_TEMPLATE_DIRECTORY = os.getenv(
```
Having a long, descriptive variable name is good. Having a short, descriptive variable name is better.
`INVOICE_UTILS_TEMPLATES_DIR` is a better name because:
- it designates a directory for all templates, not just the "body" templates
- it's shorter
- you still know what the variable refers to

koosie0507 commented on tests/test_web.py:
```diff
@@ -82,6 +83,24 @@ def email_body():
 
+@pytest.fixture
+def rendered_body_template(environment, email_invoice_request_body):
+    env = Environment(
```
I ask again: please don't put template building logic in the test code. Use a constant `str` for the expected output.

koosie0507 commented on tests/test_web.py:
```diff
@@ -199,3 +218,30 @@ def test_smtp_starttls_env_flag_must_be_true_boolean(environment, http, server,
+    ],
+    indirect=["environment"]
+)
+def test_email_body_was_sent_with_expected_body(
```
Much nicer! 👍🏼

koosie0507 commented on tests/test_web.py:
```diff
@@ -199,3 +218,30 @@ def test_smtp_starttls_env_flag_must_be_true_boolean(environment, http, server,
     assert server.starttls.call_count == 0
+
+
+@pytest.mark.parametrize(
```
Is there a test for when these env vars aren't set?

adrianmoloce replied:
I thought about it, basically the rest of the tests do this, they would go to default values and as proof the tests still pass.
If I were to test with the defaults in the same tests, it would require another expected_string and a template dir other than tests.

koosie0507 replied:
Can you write a test that fails? If you can then you should in order to increase the coverage.

adrianmoloce replied:
pytest-cov also confirmed that the entire config file is tested including default values. I had a hard time understanding what test to add here. Best I could do is add a new test which mocks jinja2.Environment and checks that it's PackageLoader and get_template params are the defaults of those env variables.

koosie0507 replied:
`pytest-cov` uses line coverage by default which is complete garbage so don't look there for answers. You can get `coverage` to do a bit better [using branch coverage](https://stackoverflow.com/questions/75719802/test-code-and-branch-coverage-simultanously-with-pytest). 

But that's not the entire story either. Branch coverage doesn't cover all cases. There's something called [mutation testing](https://medium.com/analytics-vidhya/unit-testing-in-python-mutation-testing-7a70143180d8) which is halfway decent, but too complex to use/maintain. 

And then there's TDD. Where you have to simply write a test that fails. If you can't write a test that fails, there's nothing left to cover.

koosie0507 replied:
How about making the template used in tests different from the default? Does not specifying the template related variables still yield the expected result?

adrianmoloce replied:
With the initial test, I specify the test templates which are different from the default ones and an expected_body for the test templates. I can't use the same test as I would need another expected_body for the default templates.
Not specifying the template related variables yields the intended results in all tests as they still pass.

I tried with a test where I specify wrong environment variables values  such as "bad_directory" to check that different errors like TemplateNotFound, ModuleNotFound, etc, are raised but that didn't make much sense to me so I didn't add it.

koosie0507 commented on tests/test_web.py:
```diff
@@ -199,3 +218,30 @@ def test_smtp_starttls_env_flag_must_be_true_boolean(environment, http, server,
+):
+    http.post("/invoice", json=email_invoice_request_body)
+
+    assert server.sendmail.call_args_list == [
```
This assert was already made in another test.

amoloce committed:
fix: Replace template building logic from fixture with str

  - Replaced template building from fixture with str
  - Removed a space from the test template
  - Removed redundant assert from test
  - Renamed templates directory env variable

amoloce committed:
feat: Add pytest-cov to dev dependencies

amoloce committed:
fix: Test mail body is rendered with default env values.

  - Add test to explicitly check mail body is rendered with default env
    values

amoloce committed:
fix: Test template variables with bad data.
